KAFKA-4682: Revise expiration semantics of consumer group offsets (KIP-211 - Part 1)

vahidhashemian opened the pull request:
The implementation of [expiration semantics for unsubscribed topics](https://cwiki.apache.org/confluence/display/KAFKA/KIP-211%3A+Revise+Expiration+Semantics+of+Consumer+Group+Offsets#KIP-211:ReviseExpirationSemanticsofConsumerGroupOffsets-UnsubscribingfromaTopic) will be done in a separate PR.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vahidhashemian committed:
KAFKA-4682: Revise expiration semantics of consumer group offsets (KIP-211)

vahidhashemian commented:
retest this please

vahidhashemian commented:
@hachikuji, I'd appreciate your feedback on this PR when you get a chance.

hachikuji commented on core/src/main/scala/kafka/api/ApiVersion.scala:
```diff
@@ -76,7 +76,9 @@ object ApiVersion {
-    "1.1" -> KAFKA_1_1_IV0
+    "1.1" -> KAFKA_1_1_IV0,
+    // Introduced OffsetCommitRequest V4 via KIP-211
+    "1.2-IV0" -> KAFKA_1_2_IV0
```
This should be 2.0 I think.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -171,6 +188,7 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
   private[group] val lock = new ReentrantLock
 
   private var state: GroupState = initialState
+  var currentStateTimestamp = System.currentTimeMillis()
```
Maybe we should be using a `Time` instance so that this can be mocked?

vahidhashemian replied:
Sounds good. The new commit attempts to address this.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +453,61 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+        // no consumer exists in the group =>
+        // if retention period has passed since group became Empty, expire all offsets with no pending offset commit
+        offsets.filter { case (topicPartition, commitRecordMetadataAndOffset) =>
+            !pendingOffsetCommits.contains(topicPartition) && {
```
nit: this looks misaligned

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +453,61 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+        }.toMap
+    }
+
+    System.out.println(s"${System.currentTimeMillis()} - Expired offsets from group '$groupId': ${expiredOffsets.keySet}")
```
Should this be a log message?

vahidhashemian replied:
Thanks for catching this. I used this to test the patch locally, and forgot to remove it. It could be useful as a debug message though.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +453,61 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+        // has passed since their last commit
+        val subscribedTopics = allMemberMetadata.flatMap { memberMetadata =>
+          memberMetadata.supportedProtocols.flatMap { supportedProtocol =>
+            ConsumerProtocol.deserializeSubscription(ByteBuffer.wrap(supportedProtocol._2)).topics().asScala.toSet
```
This only works if the protocol type is "consumer." I'd suggest leaving for another patch. We need to think through a more general approach for adding behavior to the coordinator which depends on the protocol type.

vahidhashemian replied:
Sounds good. Could you please clarify if any changes is required here (for this PR)?

hachikuji replied:
I would suggest that we leave subscription checking out of this patch and implement it in a follow-up. For now, maybe we can just have expiration for all offsets begin when the group becomes empty.  What I am imagining is basically a general mechanism for "installing" protocol type behavior, but it may take a few iterations to get to something that makes sense, so I'd rather not couple it with the rest of this.

vahidhashemian replied:
Ok, thanks for clarifying. This means this PR would become Part 1 of KIP-211 implementation, as we are leaving a piece of it out (i.e. no additional KIP would be required when we come back to implementing this piece). Correct me if I'm wrong.

hachikuji replied:
Yes, right.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +453,61 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+        // if retention period has passed since group became Empty, expire all offsets with no pending offset commit
+        offsets.filter { case (topicPartition, commitRecordMetadataAndOffset) =>
+            !pendingOffsetCommits.contains(topicPartition) && {
+              commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp match {
```
Maybe we could use an `Option` for the offset expiration timestamp.

vahidhashemian replied:
I refactored the code in the new commit to use an `Option` as you suggested.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java:
```diff
@@ -106,15 +106,22 @@
     public static final int DEFAULT_GENERATION_ID = -1;
     public static final String DEFAULT_MEMBER_ID = "";
     public static final long DEFAULT_RETENTION_TIME = -1L;
+    public static final long DEFAULT_EXPIRATION_TIMESTAMP = -1;
```
Is this sentinel only used in the broker code? Maybe it doesn't need to be here?

vahidhashemian replied:
Yes, currently it's being used in broker side only. So moving this definition to broker seems reasonable. I did the move in the new commit, but am not sure I put it in the best place (didn't want to change the existing class/object visibilities).

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1070,12 +1090,11 @@ object GroupMetadataManager {
   private[group] def offsetCommitValue(offsetAndMetadata: OffsetAndMetadata): Array[Byte] = {
-    // generate commit value with schema version 1
+    // generate commit value with schema version 2
     val value = new Struct(CURRENT_OFFSET_VALUE_SCHEMA)
```
nit: from the original code, but it's kind of weird to refer to the "current" schema here, but use v2 (or v1) fields below.

vahidhashemian replied:
Yes. Would you suggest removing it and use explicit `_V1` or `_V2`, or ...?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1070,12 +1090,11 @@ object GroupMetadataManager {
-    value.set(OFFSET_VALUE_OFFSET_FIELD_V1, offsetAndMetadata.offset)
-    value.set(OFFSET_VALUE_METADATA_FIELD_V1, offsetAndMetadata.metadata)
-    value.set(OFFSET_VALUE_COMMIT_TIMESTAMP_FIELD_V1, offsetAndMetadata.commitTimestamp)
-    value.set(OFFSET_VALUE_EXPIRE_TIMESTAMP_FIELD_V1, offsetAndMetadata.expireTimestamp)
```
Maybe I'm missing something, but don't we need to be able to use the old version of the offset commit schema if the expiration timestamp is provided by the user?

vahidhashemian replied:
My mistake! Updated in the new commit. Thanks!

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1211,13 +1243,16 @@ object GroupMetadataManager {
       val value = valueSchema.read(buffer)
 
-      if (version == 0 || version == 1) {
+      if (version == 0 || version == 1 || version == 2) {
```
nit: `version >= 0 && version <= 2`?

hachikuji reviewed (commented):
Thanks for the patch! Left a few initial comments.

guozhangwang commented:
@vahidhashemian are you still working on this PR?

vahidhashemian commented:
@guozhangwang I'm on vacation until next week. Will resume working on this when I return.

vahidhashemian committed:
Addressed first round of feedback

vahidhashemian reviewed (commented):
@hachikuji, thanks for the feedback and sorry for my late response. I tried to address your suggestions in the new commit (see my comments inline). Thanks!

vahidhashemian committed:
Skip the expiration semantics for unsubsribed topics for now

This section of the KIP will be implemented in a follow-up PR:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-211%3A+Revise+Expiration+Semantics+of+Consumer+Group+Offsets#KIP-211:ReviseExpirationSemanticsofConsumerGroupOffsets-UnsubscribingfromaTopic

vahidhashemian committed:
Minor updates based on feedback

vahidhashemian committed:
Minor fixes according to feedback

hachikuji commented on core/src/main/scala/kafka/api/ApiVersion.scala:
```diff
@@ -71,6 +71,7 @@ object ApiVersion {
     // and KafkaStorageException for fetch requests.
     KAFKA_1_1_IV0,
     // Introduced OffsetsForLeaderEpochRequest V1 via KIP-279
+    // Introduced OffsetCommitRequest V4 via KIP-211
```
We're obviously not consistent about this, but we really only need to mention the inter-broker protocol changes here.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java:
```diff
@@ -193,7 +199,8 @@ public OffsetCommitRequest build(short version) {
                 case 3:
-                    long retentionTime = version == 1 ? DEFAULT_RETENTION_TIME : this.retentionTime;
+                case 4:
+                    long retentionTime = version == 1 || version == 4 ? DEFAULT_RETENTION_TIME : this.retentionTime;
```
We have this ability to set the retention time, but we don't really use it anywhere in the code other than test cases. Since we're not planning on supporting it going forward, can we just remove the `retentionTime` field and just hard-code the use of `DEFAULT_RETENTION_TIME` for older requests?

vahidhashemian replied:
I'm not sure I understand your point. Shouldn't older request (v2 or v3) still be able to provide a customized retention time? Wouldn't removing the field prevent that?

hachikuji replied:
My point is that we haven't actually exposed that capability. The consumer always uses the default value. Since we don't need it going forward, we can just simplify things internally.

vahidhashemian replied:
Sounds good. I have removed this field in the new commit. Please check if the change covers what you have in mind. Thanks!

hachikuji commented on core/src/main/scala/kafka/common/OffsetMetadataAndError.scala:
```diff
@@ -28,13 +28,14 @@ case class OffsetMetadata(offset: Long, metadata: String = OffsetMetadata.NoMeta
 object OffsetMetadata {
   val InvalidOffset: Long = -1L
   val NoMetadata: String = ""
+  //val DefaultExpirationTimestamp: Long = -1
```
nit: remove this?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -112,18 +116,32 @@ private object GroupMetadata {
 
+  val DefaultCurrentStateTimestamp: Long = -1
+
   def loadGroup(groupId: String,
```
This overload feels like overkill. Can we just have the caller pass the timestamp?

vahidhashemian replied:
Sure, updated this in the new commit. If you think the default val should be defined somewhere else please let me know.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -167,10 +185,11 @@ case class CommitRecordMetadataAndOffset(appendedBatchOffset: Option[Long], offs
  */
 @nonthreadsafe
-private[group] class GroupMetadata(val groupId: String, initialState: GroupState) extends Logging {
+private[group] class GroupMetadata(val groupId: String, initialState: GroupState, time: Time = Time.SYSTEM) extends Logging {
```
We prefer not to have optional arguments (at least outside of test code) since they are too easy to overlook.

vahidhashemian replied:
I added this to facilitate mocking time in test cases when needed. Similar optional argument has been used in other places too (anywhere a `Time` instance like this is defined), e.g. in `Replica.scala`, `Log.scala`, `Throttler.scala`, ...). The actual code does not use an alternate value. Should I still change it?

hachikuji replied:
Yes, I think it is better to require all arguments be explicitly provided.

vahidhashemian replied:
Sounds good to me. I've update the code related to this class definition.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -657,10 +657,10 @@ class GroupMetadataManager(brokerId: Int,
-        offsetAndMetadata.copy(expireTimestamp = offsetAndMetadata.commitTimestamp + config.offsetsRetentionMs)
-      else
-        offsetAndMetadata
+        if (offsetAndMetadata.expireTimestamp.contains(org.apache.kafka.common.requests.OffsetCommitRequest.DEFAULT_TIMESTAMP))
```
nit: `OffsetCommitRequest` is imported above

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -657,10 +657,10 @@ class GroupMetadataManager(brokerId: Int,
-      else
-        offsetAndMetadata
+        if (offsetAndMetadata.expireTimestamp.contains(org.apache.kafka.common.requests.OffsetCommitRequest.DEFAULT_TIMESTAMP))
+          offsetAndMetadata.copy(expireTimestamp = Some(offsetAndMetadata.commitTimestamp + config.offsetsRetentionMs))
```
Is there a strong reason to use copy rather than just constructing a new object?

vahidhashemian replied:
This is not my addition. Isn't this more concise than the alternative?
```
          new OffsetAndMetadata(offsetAndMetadata.offsetMetadata,
            offsetAndMetadata.commitTimestamp, 
            expireTimestamp = Some(offsetAndMetadata.commitTimestamp + config.offsetsRetentionMs))
```
I can construct a new object if you think that's better practice.

hachikuji replied:
Probably not that big of a deal, but if we added any additional fields, it would force us to consider whether we wanted to copy it in this context.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -353,10 +354,12 @@ class KafkaApis(val requestChannel: RequestChannel,
-                defaultExpireTimestamp
+              if (header.apiVersion >= 3)
+                None
+              else  if (partitionData.timestamp == OffsetCommitRequest.DEFAULT_TIMESTAMP)
```
nit: extra space between else and if

hachikuji reviewed (commented):
Thanks for the updates, left a few more comments.

vahidhashemian reviewed (commented):
@hachikuji Addressed some of your suggestions in the new commits. Would appreciate a couple of clarifications inline. Thanks!

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +441,54 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+        }.toMap
+
+      case Some(protocol) =>
+        //TODO (KIP-211): expire offsets of partitions the group is no longer subscribed to, assuming retention period
```
We can post a JIRA to keep track of this rather than doing it in the code. For now, maybe we can just remove this case?

vahidhashemian replied:
Sure, it looks like a `case _ => Map()` is required at the end to cover all cases.
I created [KAFKA-6951](https://issues.apache.org/jira/browse/KAFKA-6951) to keep track of this portion.

vahidhashemian commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -657,10 +657,12 @@ class GroupMetadataManager(brokerId: Int,
+        if (offsetAndMetadata.expireTimestamp.contains(OffsetCommitRequest.DEFAULT_TIMESTAMP))
+          new OffsetAndMetadata(offsetAndMetadata.offsetMetadata,
+            offsetAndMetadata.commitTimestamp,
+            expireTimestamp = Some(offsetAndMetadata.commitTimestamp + config.offsetsRetentionMs))
```
I replaced the `copy(...)` with direct object construction in the new commmit.

vahidhashemian reviewed (commented):
@hachikuji, thanks again for the recent feedback. They should be addressed in the new commit. Please advise if anything is missing.

hachikuji commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -140,9 +140,9 @@ public void testSerialization() throws Exception {
-        checkRequest(createOffsetCommitRequest(2));
-        checkErrorResponse(createOffsetCommitRequest(2), new UnknownServerException());
-        checkResponse(createOffsetCommitResponse(), 0);
+        checkRequest(createOffsetCommitRequest(4));
```
nit: maybe we can move this to where the rest of the offset commit requests are being tested. This test case is as vexing as ever 😞.

vahidhashemian committed:
Updated upgrade notes

Removed the sentence about changing how offset expiration for unsubscribed topic works, as it will be implemented in a future PR.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -125,7 +125,7 @@ class GroupCoordinator(val brokerId: Int,
             responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
           } else {
-            val group = groupManager.addGroup(new GroupMetadata(groupId, initialState = Empty))
+            val group = groupManager.addGroup(new GroupMetadata(groupId, Empty, Time.SYSTEM))
```
GroupCoordinator already has a `Time` field.

vahidhashemian replied:
True, sorry I did not check for existence of such `Time` fields in the last commit.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1225,13 +1275,16 @@ object GroupMetadataManager {
         val leaderId = value.get(LEADER_KEY).asInstanceOf[String]
         val memberMetadataArray = value.getArray(MEMBERS_KEY)
         val initialState = if (memberMetadataArray.isEmpty) Empty else Stable
+        val currentStateTimestamp: Long =
```
I wonder if it would be safer to use an `Option` for this. Then the type checker can help us ensure that it is used correctly.

vahidhashemian replied:
I made the switch in the new commit.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1247,7 +1300,7 @@ object GroupMetadataManager {
-        GroupMetadata.loadGroup(groupId, initialState, generationId, protocolType, protocol, leaderId, members)
+        GroupMetadata.loadGroup(groupId, initialState, generationId, protocolType, protocol, leaderId, currentStateTimestamp, members)
       } else {
         throw new IllegalStateException("Unknown group metadata message version")
```
Can we add the version to this message? Same for the similar case above loading offsets.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -711,11 +713,11 @@ class GroupMetadataManager(brokerId: Int,
   private[group] def cleanupGroupMetadata(): Unit = {
-    val startMs = time.milliseconds()
+    val currentTimestamp = time.milliseconds()
     val offsetsRemoved = cleanupGroupMetadata(groupMetadataCache.values, group => {
```
nit: maybe we can rename to `numOffsetsRemoved`?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -630,7 +630,7 @@ class GroupMetadataManager(brokerId: Int,
         // metadata stored in the log
         (emptyGroupOffsets.keySet ++ pendingEmptyGroupOffsets.keySet).foreach { groupId =>
-          val group = new GroupMetadata(groupId, initialState = Empty)
+          val group = new GroupMetadata(groupId, Empty, Time.SYSTEM)
```
GroupMetadataManager also has a `Time` field.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1084,14 +1106,29 @@ object GroupMetadataManager {
+    val value = new Struct(schemaForOffset(offsetValueSchemaVersion))
+
+    if (offsetValueSchemaVersion == CURRENT_OFFSET_VALUE_SCHEMA_VERSION) {
+      value.set(OFFSET_VALUE_OFFSET_FIELD_V2, offsetAndMetadata.offset)
```
I may have had a comment about this already, but isn't it weird to assume v2 if the schema version matches `CURRENT_OFFSET_VALUE_SCHEMA_VERSION`? Perhaps we should get rid of `CURRENT_OFFSET_VALUE_SCHEMA_VERSION` and use v2 directly?

vahidhashemian replied:
I took a stab at this in the new commit. Please advise if it's not what you had in mind (I feel there is room for improvement there). One benefit of `CURRENT_OFFSET_VALUE_SCHEMA_VERSION` was it prevented some hardcoding of version numbers.

hachikuji replied:
Yeah I get the intent, but our execution was not great. Even with the update, we use `CURRENT_OFFSET_VALUE_SCHEMA_VERSION` initially, but then go on and assume that it must be version 2. We don't update this schema often, so I guess I'd rather just make it explicit.

vahidhashemian replied:
I take it from your comment that we should get rid of `CURRENT_OFFSET_VALUE_SCHEMA_VERSION`altogether. I've done so in the new commit. In that case, should we do the same for other `CURRENT_...` definitions?

hachikuji commented on core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala:
```diff
@@ -506,7 +512,7 @@ class GroupMetadataManagerTest {
     assertTrue(groupMetadataManager.groupNotExists(groupId))
 
-    val group = new GroupMetadata(groupId, initialState = Empty)
+    val group = new GroupMetadata(groupId, Empty, Time.SYSTEM)
```
Is it intentional not to use the `MockTime` instance?

vahidhashemian replied:
No, sorry for missing this.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +441,51 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+        // if retention period has passed since group became Empty, expire all offsets with no pending offset commit
+        offsets.filter {
+          case (topicPartition, commitRecordMetadataAndOffset) =>
+            !pendingOffsetCommits.contains(topicPartition) && {
```
I'm puzzling a little over why we only have this check in the Empty case. Is there a reason? 

Given how similar the logic between these two branches is, I wonder if we can factor out the common structure and use a function argument for the slight behavioral difference.

vahidhashemian replied:
The check seems to be applicable to the standalone consumer too. I've added that, and also refactored the code as you suggested.

hachikuji reviewed (commented):
Thanks, left a few more comments.

vahidhashemian committed:
Minor updates to address recent feedback

vahidhashemian reviewed (commented):
@hachikuji, I pushed another commit to address your recent feedback. Thanks for the reviews.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -118,12 +119,14 @@ private object GroupMetadata {
                 protocolType: String,
                 protocol: String,
                 leaderId: String,
+                timestamp: Option[Long],
```
nit: can we name this `currentStateTimesetamp` for consistency?

hachikuji commented on core/src/main/scala/kafka/common/OffsetMetadataAndError.scala:
```diff
@@ -44,7 +44,7 @@ case class OffsetAndMetadata(offsetMetadata: OffsetMetadata,
 
 object OffsetAndMetadata {
-  def apply(offset: Long, metadata: String, commitTimestamp: Long, expireTimestamp: Long) = new OffsetAndMetadata(OffsetMetadata(offset, metadata), commitTimestamp, expireTimestamp)
+  def apply(offset: Long, metadata: String, commitTimestamp: Long, expireTimestamp: Long) = new OffsetAndMetadata(OffsetMetadata(offset, metadata), commitTimestamp, Some(expireTimestamp))
```
Hmm.. Should we be checking for the default timestamp here? I had expected that we would use `None` if the default is used.

vahidhashemian replied:
Wouldn't the default be when the `expireTimestamp` argument is not provided? which will be taken care of by the following methods?

hachikuji replied:
I had assumed so initially, but then I saw the `contains(DEFAULT_TIMESTAMP)` check below.

hachikuji commented on core/src/main/scala/kafka/common/OffsetMetadataAndError.scala:
```diff
@@ -34,7 +34,7 @@ object OffsetMetadata {
 case class OffsetAndMetadata(offsetMetadata: OffsetMetadata,
                              commitTimestamp: Long = org.apache.kafka.common.requests.OffsetCommitRequest.DEFAULT_TIMESTAMP,
-                             expireTimestamp: Long = org.apache.kafka.common.requests.OffsetCommitRequest.DEFAULT_TIMESTAMP) {
+                             expireTimestamp: Option[Long] = None) {
```
We should update the `toString` below to use string interpolation. Currently it is using `%d` incorrectly.

vahidhashemian replied:
Thanks for catching this, updated in the new commit.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +440,46 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+    }
+
+    val expiredOffsets: Map[TopicPartition, CommitRecordMetadataAndOffset] = protocolType match {
+      case Some(protocol) if is(Empty) =>
```
nit: since we don't care about the protocol (yet), we can use `_` for the name

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +440,46 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+      case Some(protocol) if is(Empty) =>
+        // no consumer exists in the group =>
+        // if retention period has passed since group became Empty, expire all offsets with no pending offset commit
+        getExpiredOffsets(commitRecordMetadataAndOffset => {currentStateTimestamp.get})
```
nit: you can simplify 
```scala
        getExpiredOffsets(_ => currentStateTimestamp.get)
```
Also, are you sure the `get` is safe? What happens for a group that is loaded from an old version of the metadata schema since it has no current timestamp? It seems like we would load `None` in that case.

vahidhashemian replied:
Thanks for the code update suggestion. You're right about the `None` issue. I switched to using `currentStateTimestampOrDefault` to avoid it.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +440,46 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+      case None =>
+        // protocolType is None => standalone (simple) consumer, that uses Kafka for offset storage only
+        // expire offsets with no pending offset commit that retention period has passed since their last commit
+        getExpiredOffsets(commitRecordMetadataAndOffset => {commitRecordMetadataAndOffset.offsetAndMetadata.commitTimestamp})
```
nit:
```scala
        getExpiredOffsets(_.offsetAndMetadata.commitTimestamp)
```

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +440,46 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+        Map()
+    }
+
+    debug(s"Expired offsets from group '$groupId': ${expiredOffsets.keySet}")
```
Can we skip this message if `expiredOffsets` is empty?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +440,46 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+
     offsets --= expiredOffsets.keySet
-    expiredOffsets.toMap
+    expiredOffsets.map {
```
Could we move this `map` into `getExpiredOffsets`?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -657,10 +657,12 @@ class GroupMetadataManager(brokerId: Int,
-        offsetAndMetadata.copy(expireTimestamp = offsetAndMetadata.commitTimestamp + config.offsetsRetentionMs)
-      else
-        offsetAndMetadata
+        if (offsetAndMetadata.expireTimestamp.contains(OffsetCommitRequest.DEFAULT_TIMESTAMP))
```
Related to my previous comment. It would be better if we don't allow both `None` and `Some(-1)`. We should be able to use `None` for both cases.

vahidhashemian replied:
Thanks for the earlier pointer that made me look more closely at this. Though, I'm having a difficult time finding out where we store `-1` for `expireTimestamp` that justifies this check, and having no luck. Any ideas? I'm thinking if a `-1` has been stored (as default) for expiration timestamp, wouldn't we break backward compatibility if we switch this default to `None`?

hachikuji replied:
Can you explain why we are doing the `contains` check here if the case is not possible? Should we be using `offsetAndMetadata.expireTimestamp.isEmpty` instead?

hachikuji replied:
Unless I'm mistaken, I think we can just remove this check and use `offsetAndMetadata`. We shouldn't need the special case for version 0 anymore because it can be treated the same as the latest version. Both use `None` as the expiration timestamp.

vahidhashemian replied:
Yes, if we want to treat version 0 similar to the current version this suggestion should work. Thanks!
It would simplify the code a bit too.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1108,13 +1145,20 @@ object GroupMetadataManager {
   private[group] def groupMetadataValue(groupMetadata: GroupMetadata,
                                         assignment: Map[String, Array[Byte]],
                                         version: Short = 0): Array[Byte] = {
```
nit: can we remove the optional value?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -39,7 +39,7 @@ import org.apache.kafka.common.protocol.Errors
 import org.apache.kafka.common.protocol.types._
 import org.apache.kafka.common.record._
-import org.apache.kafka.common.requests.{IsolationLevel, OffsetFetchResponse}
+import org.apache.kafka.common.requests.{IsolationLevel, OffsetCommitRequest, OffsetFetchResponse}
```
Putting this comment here for lack of a better place. We need to update `storeGroup` to check the inter-broker format before using the new version of the group metadata schema. I am puzzled why we don't have a similar check for `offsetCommitValue`. I suspect it is a bug. We could fix it here or in a follow-up.

vahidhashemian replied:
Thanks for pointing this out. I fixed it in the new commit. Regarding `offsetCommitValue`, wouldn't the existing check in `offsetCommitValue(...)` handle the expected version detection?

    val offsetValueSchemaVersion: Short =
      // if expire timestamp is explicitly provided use the old schema
      if (offsetAndMetadata.expireTimestamp.nonEmpty) 1
      //  otherwise, use the current v2 schema
      else 2

hachikuji replied:
Hmm.. Not sure how that would be sufficient. The point is to protect the rolling upgrade scenario. So if the leader upgrades first and begins using the new schema version, then a follower which has not yet upgraded may fail if something causes it to become leader. This check only ensures that we use the version needed to implement the expected client semantics.

vahidhashemian replied:
Thanks for the clarification. I've added something similar for `offsetCommitValue` (with minor refactoring) that supposedly replaces the block of code in my previous post.

hachikuji reviewed (commented):
Sorry for the delay. Left a few more comments.

vahidhashemian committed:
Addressed recent feedback

vahidhashemian reviewed (commented):
@hachikuji, thanks for the recent feedback. The new commit should cover them for the most part. A bit confused about the `-1` default value for expiration timestamp as indicated earlier. Also, a question inline on inter-broker format check for `offsetCommitValue`. Thanks!

hachikuji commented:
@vahidhashemian Can you rebase please?

vahidhashemian commented:
@hachikuji, sorry for the delay. Rebased the PR.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -354,10 +355,12 @@ class KafkaApis(val requestChannel: RequestChannel,
+              if (header.apiVersion >= 3)
+                None
+              else if (partitionData.timestamp == OffsetCommitRequest.DEFAULT_TIMESTAMP)
+                Some(defaultExpireTimestamp)
```
Can we use `None` here? I thought the idea in the KIP was to treat older version of the offset commit which have the default timestamp set under the new expiration logic.

hachikuji commented on core/src/main/scala/kafka/common/OffsetMetadataAndError.scala:
```diff
@@ -34,17 +34,24 @@ object OffsetMetadata {
 
-  override def toString = "[%s,CommitTime %d,ExpirationTime %d]".format(offsetMetadata, commitTimestamp, expireTimestamp)
+  override def toString = {
+    "[%s,CommitTime %d".format(offsetMetadata, commitTimestamp) + {
```
Can we not use string interpolation?
```scala
s"[$offsetMetadata, CommitTime $commitTimestamp, ExpirationTime $expireTimestamp]"
```

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -118,12 +119,14 @@ private object GroupMetadata {
+                currentStateTimestamp: Option[Long],
                 members: Iterable[MemberMetadata]): GroupMetadata = {
-    val group = new GroupMetadata(groupId, initialState)
+    val group = new GroupMetadata(groupId, initialState, Time.SYSTEM)
```
We should avoid hard-coding references to `Time.SYSTEM` when possible. Maybe we can pass the time instance as another argument?

vahidhashemian replied:
Sure, but it seems the hard-coding is inevitable (at least without major refactoring) in some places. I'll let you review those in the recent commit.

hachikuji commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -488,6 +488,10 @@ object ConsoleConsumer extends Logging {
       case None =>
         consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, s"console-consumer-${new Random().nextInt(100000)}")
+        // By default, avoid unnecessary expansion of the coordinator cache since
+        // the auto-generated group and its offsets won't likely be used again
```
nit: change "won't likely" to "is not intended to"?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -434,18 +440,47 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+            commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp match {
+              case None =>
+                // current version with no per partition retention
+                currentTimestamp - baseTimestamp(commitRecordMetadataAndOffset) >= offsetRetentionMs
```
This is a little confusing. The function `baseTimestamp` can return -1 if no current state timestamp is set, which would cause this check to incorrectly succeed. This would be the case if we were loading the state of a group which existed before we began using the new group metadata schema. Are there any guards which prevent this case?

Basically the problem is that we need to handle the case that there is no expiration timestamp and there is no current state timestamp. I think in this case, `baseTimestamp` should just use the commit timestamp, which is basically like reverting to the old behavior. In other words, we should only be allowed to use the new behavior if `currentStateTimestamp` is defined.

vahidhashemian replied:
That's correct. Your suggested solution should work, and is implemented in the new commit. Thanks!

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1225,13 +1277,16 @@ object GroupMetadataManager {
         val memberMetadataArray = value.getArray(MEMBERS_KEY)
         val initialState = if (memberMetadataArray.isEmpty) Empty else Stable
+        val currentStateTimestamp: Option[Long] =
+          if (version > 1) Some(value.getLong(CURRENT_STATE_TIMESTAMP_KEY))
```
When we are serializing the value, we use -1 if `currentStateTimestamp` is not defined. Shouldn't we have a matching check here so that -1 is converted to `None`?

hachikuji commented:
@vahidhashemian Thanks for the updates. I left a few more comments. This patch is turning out trickier than I imagined and I'm a little troubled that we don't have test cases covering the issues I'm finding. As of this second, I'm leaning toward pushing this to the next release, but let's wait on that decision until the pending comments have been addressed.

vahidhashemian committed:
Addressed recent feedback

vahidhashemian reviewed (commented):
@hachikuji, thanks for another round of review and pointing out some of my misses. It's perfectly fine if you decide to move this PR to the next release.

New commit tries to address your recent feedback. Additional responses are inline.

hachikuji commented on core/src/main/scala/kafka/common/OffsetMetadataAndError.scala:
```diff
@@ -41,9 +41,9 @@ case class OffsetAndMetadata(offsetMetadata: OffsetMetadata,
 
   override def toString = {
-    "[%s,CommitTime %d".format(offsetMetadata, commitTimestamp) + {
+    s"[$offsetMetadata,CommitTime $commitTimestamp" + {
```
nit: I think it's fine to leave `expireTimestamp` formatted as an `Option` to simplify this

vahidhashemian replied:
👍

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1105,32 +1084,33 @@ object GroupMetadataManager {
+                                       apiVersion: ApiVersion): Array[Byte] = {
     // generate commit value according to schema version
+    val (version, value) = {
+      if (apiVersion < KAFKA_2_0_IV2)
```
I think I'm missing something again. If we are dealing with an old OffsetCommit which contains a custom retention time, don't we need to use v1 also. Do we have a test case for this scenario?

vahidhashemian replied:
Isn't it what the `if` block actually does? i.e. if the API version corresponds to an older version use version 1 of the offset value schema. In the new commit I added a test case for loading offsets from old commits that runs through this check. Would that suffice as a test case for this part?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1284,9 +1269,16 @@ object GroupMetadataManager {
-        val currentStateTimestamp: Option[Long] =
-          if (version > 1) Some(value.getLong(CURRENT_STATE_TIMESTAMP_KEY))
-          else None
+        val rawCurrentStateTimestamp = value.getLong(CURRENT_STATE_TIMESTAMP_KEY)
```
Won't this raise an exception for the older schemas? Do we have any test cases covering this?

vahidhashemian replied:
Yes, fixed it in the new commit, and should be covered in the new unit test.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1284,9 +1269,16 @@ object GroupMetadataManager {
-          else None
+        val rawCurrentStateTimestamp = value.getLong(CURRENT_STATE_TIMESTAMP_KEY)
+        val currentStateTimestamp: Option[Long] = version match {
+          case version if version >= 2 =>
```
We know from the check above that `version <= 2`, so can we just check for `2`?

vahidhashemian replied:
I was thinking if a version 3 is added later on this check would keep working (assuming we're not undoing this KIP). Anyway, I've changed it in the new commit.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -355,10 +355,8 @@ class KafkaApis(val requestChannel: RequestChannel,
             commitTimestamp = currentTimestamp,
             expireTimestamp = {
-              if (header.apiVersion >= 3)
+              if (header.apiVersion >= 2 || partitionData.timestamp == OffsetCommitRequest.DEFAULT_TIMESTAMP)
```
I am not sure about this check. I think we should use `None` for every case except when the version is v2 and the retention time is set to use the default. In fact, can't we just do this?
```
val retentionTime = offsetCommitRequest.retentionTime
if (retentionTime == DEFAULT_RETENTION_TIME)
  Some(currentTimestamp + retentionTime)
else
  None
```
Only v2 requests will have anything other than `DEFAULT_RETENTION_TIME`. For everything else, we'll use the new expiration semantics.

vahidhashemian replied:
For v2/v3/v4 the retention time could be something other than the default.
For v1 the commit time could be something other than the default, in which case we calculate an expiration timestamp.
The expiration time would be `None` if retention time is the default.
```
            expireTimestamp = offsetCommitRequest.retentionTime match {
              case OffsetCommitRequest.DEFAULT_RETENTION_TIME => None
              case retentionTime =>
                partitionData.timestamp match {
                  case OffsetCommitRequest.DEFAULT_TIMESTAMP => None
                  case commitTimestamp => Some(commitTimestamp + retentionTime)
                }
            }
```
I used the above for expiration timestamp. Let me know if that works.

hachikuji reviewed (commented):
Thanks for the update. Left a few more comments.

vahidhashemian committed:
Backward compatibility improvements

vahidhashemian reviewed (commented):
@hachikuji, thanks for the feedback and identifying those issues. I tried to address them in the new commit. Additional comments inline.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -331,33 +331,29 @@ class KafkaApis(val requestChannel: RequestChannel,
+            expireTimestamp = offsetCommitRequest.retentionTime match {
+              case OffsetCommitRequest.DEFAULT_RETENTION_TIME => None
+              case retentionTime =>
+                partitionData.timestamp match {
```
My thought was that we could just ignore the commit timestamp and treat that as an instance of the new semantics. This simplifies the code and is probably what users would actually want. Since only the commit timestamp is provided for these old versions, there wouldn't be any client guarantees on retention anyway.

vahidhashemian replied:
That should work. I see a minor risk that the new behavior would not fully comply with what the users of the old versions expect. Perhaps a far fetched example would be using an explicit commit timestamp in the future and expecting the message stay longer than what the new semantics implements.

In any case, I updated this in the new commit.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -1081,17 +1083,33 @@ object GroupMetadataManager {
+                                       apiVersion: ApiVersion): Array[Byte] = {
+    // generate commit value according to schema version
+    val (version, value) = {
+      if (apiVersion < KAFKA_2_0_IV2)
```
Continuing from previous comment. This uses the inter-broker version to determine the offset commit schema, but we also have to consider the version of the OffsetCommit request, right? If a user has requested an explicit retention time, then we have to use the old version because the new schema does not have an expiration time. Maybe I have misunderstood something?

vahidhashemian replied:
No, that seems to be correct. Hopefully, the new commit covers this. Thanks!

vahidhashemian committed:
Minor change in the expiration semantics with respect to old versions

vahidhashemian reviewed (commented):
@hachikuji, thanks for review. And I thought I did a decent job with the KIP :) Backward compatibility is a beast! Please check the new commit when you have a chance. Thanks!

vahidhashemian commented:
retest this please

hachikuji commented:
@vahidhashemian Yeah, compatibility is tough. I think we are also hitting some gaps in the existing tests. Anyway, I'll do another pass on Monday and hopefully we'll be able to merge soon.

hachikuji commented on core/src/main/scala/kafka/common/OffsetMetadataAndError.scala:
```diff
@@ -34,17 +34,17 @@ object OffsetMetadata {
   def metadata = offsetMetadata.metadata
 
-  override def toString = "[%s,CommitTime %d,ExpirationTime %d]".format(offsetMetadata, commitTimestamp, expireTimestamp)
+  override def toString = s"[$offsetMetadata,CommitTime $commitTimestamp,ExpirationTime ${expireTimestamp.getOrElse("_")}"
```
nit: missing end bracket.

hachikuji commented on core/src/main/scala/kafka/api/ApiVersion.scala:
```diff
@@ -73,7 +73,8 @@ object ApiVersion {
     // Introduced ApiVersionsRequest V2 via KIP-219
-    KAFKA_2_0_IV1
+    KAFKA_2_0_IV1,
+    KAFKA_2_0_IV2
```
I think it's legitimate to bump the internal broker version for the schema change for offset/group metadata. Perhaps we can mention this here? Also, we should probably use create a new version for Kafka 2.1.

hachikuji reviewed (approved):
@vahidhashemian Sorry for the delay. LGTM. We just need to update the version references and the upgrade notes. I had one small nitpick as well.

vahidhashemian committed:
Bumped the inter-broker version

vahidhashemian commented:
@hachikuji thanks for taking another look. I made the changes you suggested.

hachikuji commented:
@vahidhashemian Just one more thing. We need to fix the upgrade notes so that they list KIP-211 for version 2.1.

vahidhashemian committed:
Updated the target release version

vahidhashemian commented:
@hachikuji thanks for catching that. I made changes in a few places to update the target release version to 2.1.

vahidhashemian commented:
retest this please

hachikuji commented:
@vahidhashemian Thanks for the patch. I think the config description for `offsets.retention.minutes` needs to be updated as well. Would you mind submitting a follow-up?

vahidhashemian commented:
@hachikuji Sure, will submit one within the next few hours. Thanks!

ijuma commented on docs/upgrade.html:
```diff
@@ -66,6 +78,9 @@ <h4><a id="upgrade_2_0_0" href="#upgrade_2_0_0">Upgrading from 0.8.x, 0.9.x, 0.1
     </li>
+    <li>The default for console consumer's <code>enable.auto.commit</code> property when no <code>group.id</code> is provided is now set to <code>false</code>.
+        This is to avoid polluting the consumer coordinator cache as the auto-generated group is not likely to be used by other consumers.
+    </li>
```
@vahidhashemian Was this included in 2.0?

vahidhashemian replied:
@ijuma no, sorry, I missed moving that one. Will update this in the follow-up PR shortly.

aldex32 commented:
Hi guys,

will this fix be backported in older Kafka versions?

vahidhashemian commented:
@aldex32 I don't think there is a plan for that. @ijuma or @hachikuji can confirm.

hachikuji commented:
We are typically very conservative about the compatibility of older releases. This KIP is borderline from a compatibility perspective, but it did require a protocol bump, so we are unlikely to backport.

aldex32 commented:
Thanks for the answer guys.
